### PR TITLE
Simple GpuProfiler implementation for WebGL and WebGPU

### PR DESCRIPTION
**Pull request by mvaligursky**
Fixes https://github.com/playcanvas/engine/issues/4834

Implementation of a GPU Profiler:
- on WebGL 2, a frame duration is captured using `gl.beginQuery` and `gl.endQuery`. Ideally `gl.queryCounterEXT` would be used to easily allow render pass timing as well, but that is not available on any platform. Render pass timing can be simulated at a later stage using the begin/end queries, which unfortunately do not support nesting.
- on WebGPU a frame duration as well render pass durations are captured.
- profiling needs to be enabled, otherwise it's not active. Additionally, a tracing can be enabled:
Screenshot 2023-07-11 at 13 17 52

and this is the output on WebGPU. Webgl only outputs the first line:

Screenshot 2023-07-11 at 13 16 35

New public API

on GraphicsDevice:

gfx

new class:

profiler

Additionally, the MiniStats were updated to obtain the GPU frame duration from profiler, and its internal WebGl code has been removed.
Note that MiniStats are still not supported on WebGPU due to further refactoring required.

Extensions: https://registry.khronos.org/webgl/extensions/EXT_disjoint_timer_query_webgl2/

### CODE REVIEW

File: src/platform/graphics/gpu-profiler.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import { TRACEID_GPU_TIMINGS } from "../../core/constants.js";
+import { Debug } from "../../core/debug.js";
+import { Tracing } from "../../core/tracing.js";
+
+/**
+ * A class implementing a simple GPU profiler.
```
[the remark this answers is not in the document]
Comment (slimbuck) [reply src/platform/graphics/gpu-profiler.js:6]: TBH I find this base class and the inherited class usage pretty confusing. What is this base class tracking? How does it get used?

Even though these are internal details, perhaps adding some comments will help.

File: src/platform/graphics/webgpu/webgpu-graphics-device.js [hunk 2]
```diff
@@ -283,6 +283,8 @@ class WebgpuGraphicsDevice extends GraphicsDevice {
     postInit() {
         super.postInit();
 
+        this.gpuProfiler = new WebgpuGpuProfiler(this);
```
Comment (slimbuck) [inline src/platform/graphics/webgpu/webgpu-graphics-device.js:286]: Should profiler always be created? For one thing this'll mean we cannot tree shake it.

**Comment by slimbuck** [comment]
The difference is that GpuProfiler (and the corresponding webgl/gpu bits) is never needed for production app, whereas Texture is.

MiniStats was placed in extras so everyone doesn't have to take the hit.

**Comment by mvaligursky** [comment]
I see the future where GPU timing is available for production app, as many other engines expose.

But regardless, at the moment we do not have a system where we can strip per platform implementation classes, apart from exposing those to the user to allocate. All per platforms classes are always imported if the relevant device class, which is a factory of those classes, is imported.

**Review by slimbuck — APPROVED** [review]

**Pull request merged**